packages/loopring_v3/contracts/amm/libamm/AmmExitProcess.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../aux/transactions/TransactionReader.sol";
+import "../../core/impl/libtransactions/TransferTransaction.sol";
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/MathUint96.sol";
+import "../../thirdparty/SafeCast.sol";
+import "./AmmUtil.sol";
+import "./AmmData.sol";
+import "./AmmExitRequest.sol";
+import "./AmmPoolToken.sol";
+import "./AmmStatus.sol";
+
+
+/// @title AmmExitProcess
+library AmmExitProcess
+{
+    using AmmPoolToken      for AmmData.State;
+    using AmmStatus         for AmmData.State;
+    using AmmUtil           for uint96;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+    using TransactionReader for ExchangeData.Block;
+
+    function proxcessExchangeWithdrawal(
+        AmmData.State    storage S,
+        AmmData.Context  memory  ctx,
+        AmmData.Token    memory  token,
+        uint                     amount
+        )
+        internal
+    {
+        // Check that the withdrawal in the block matches the expected withdrawal
+        WithdrawTransaction.Withdrawal memory withdrawal = ctx._block.readWithdrawal(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        // These fields are not read by readWithdrawal:
+        withdrawal.minGas = 0;
+        withdrawal.to = address(this);
+        withdrawal.extraData = new bytes(0);
+
+        require(
+            withdrawal.withdrawalType == 1 && // Question(brecht): should this be 1?
+            withdrawal.owner == address(this) &&
+            withdrawal.accountID == S.accountID &&
+            withdrawal.tokenID == token.tokenID &&
+            // withdrawal.amount.isAlmostEqual(amount) && // Question(brecht):should we use isAlmostEqual?
+            withdrawal.amount == amount &&
+            withdrawal.feeTokenID == 0 &&
+            withdrawal.fee == 0 &&
+            withdrawal.storageID == 0 && // Question(brecht):should we check this ID?
```

Comment:
No, the relayer can freely set the storegeID to any value that makes the withdrawal valid. We don't know what value that is onchain (storageID == 0 can only be used once).